src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/UseStringContainsCharOverloadWithSingleCharactersAnalyzer.cs

```diff
@@ -0,0 +1,87 @@
+            CA1839,
+            s_localizableTitle_CA1839,
+            s_localizableMessage_CA1839,
+            DiagnosticCategory.Performance,
+            RuleLevel.IdeSuggestion,
+            description: s_localizableDescription_CA1839,
+            isPortedFxCopRule: false,
+            isDataflowRule: false);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(s_rule_CA1839);
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+            context.EnableConcurrentExecution();
+            context.RegisterCompilationStartAction(CheckIfRuleIsApplicableAndRegister);
+        }
+
+        private void CheckIfRuleIsApplicableAndRegister(CompilationStartAnalysisContext ctx)
+        {
+            var stringType = ctx.Compilation.GetSpecialType(SpecialType.System_String);
+            var charType = ctx.Compilation.GetSpecialType(SpecialType.System_Char);
+
+            // Bail out of further evaluations in special cases where string/char types arent even available
+            if (stringType is null || charType is null)
+                return;
+
+            var ruleIsApplicable = stringType.GetMembers("Contains").OfType<IMethodSymbol>().Any(m =>
+            {
+                if (m.Parameters.Length > 0)
+                {
+                    return SymbolEqualityComparer.Default.Equals(m.Parameters[0].Type, charType);
+                }
+                return false;
+            });
+
+            if (ruleIsApplicable)
+                ctx.RegisterOperationAction(AnalyseContainsUsage, OperationKind.Invocation);
+        }
+
+        private void AnalyseContainsUsage(OperationAnalysisContext ctx)
+        {
+            var invocationOperation = (IInvocationOperation)ctx.Operation;
+            if (ShouldAnalyzeInvocations(invocationOperation) && ViolatesRule(invocationOperation.Arguments[0]))
+            {
+                ctx.ReportDiagnostic(invocationOperation.Arguments[0].CreateDiagnostic(s_rule_CA1839));
+            }
+
+            static bool ShouldAnalyzeInvocations(IInvocationOperation invocationOperation)
+            {
+                return invocationOperation.TargetMethod is IMethodSymbol invokedMethod
+                                && invokedMethod.ContainingType.SpecialType.Equals(SpecialType.System_String)
+                                && invokedMethod.Name == "Contains"
+                                && invokedMethod.Parameters[0].Type.SpecialType.Equals(SpecialType.System_String);
+            }
+
+            static bool ViolatesRule(IArgumentOperation primaryArgument)
+            {
+                return primaryArgument.Value is ILiteralOperation literalOperation
+                    && literalOperation.ConstantValue.HasValue
+                    && literalOperation.ConstantValue.Value is string constantString
```

Comment:
💭 I believe `GetValueOrDefault()` will work here as well

```suggestion
                    && literalOperation.ConstantValue.GetValueOrDefault() is string constantString
```

Reply:
`IOperation.ConstantValue` sadly is `Optional<object>`  instead of a nullable object, so no `GetValueOrDefault` for me if theres no established helper that i dont know of 😄